Guard array access in woocommerce_subcategory_thumbnail

christopher-silvey opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
- Guard the result of `wp_get_attachment_image_src()` in `woocommerce_subcategory_thumbnail()` before reading index `0`.
- When the attachment lookup fails or returns an unexpected value, fall back to `wc_placeholder_img_src()` and set `$image_srcset` and `$image_sizes` to `false` so the existing `if ( $image )` block renders the placeholder image instead.
- Prevent the PHP warning `Trying to access array offset on value of type bool` when the category thumbnail attachment is missing or cannot be loaded.

Guard access to the image URL in `woocommerce_subcategory_thumbnail()` to avoid
`Trying to access array offset on value of type bool` warnings.

`wp_get_attachment_image_src()` can return `false` when the attachment lookup fails.
The previous code assumed the return value was always an array and unconditionally
accessed `$image[0]`, which triggers a warning when it is `false`.

This change checks that the result is an array and contains index `0` before
reading it. If that check fails, the code now uses the WooCommerce placeholder
URL from `wc_placeholder_img_src()` and disables `srcset` and `sizes`. This avoids
the warning while still rendering a predictable thumbnail.

Behavior for valid image data is unchanged.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create or identify a product category that currently has a thumbnail image set.
2. Note the attachment ID used for that category thumbnail.
3. Delete that attachment from the Media Library so the term meta still points to the old ID but the file no longer exists.
4. Visit the product category archive page for that term on the front end.
5. **On current WooCommerce without this patch**, observe the PHP error log and verify that warnings like  
   `PHP Warning:  Trying to access array offset on value of type bool in wp-content/plugins/woocommerce/includes/wc-template-functions.php on line 2844`  
   are generated.
6. Apply this patch and reload the same product category archive page.
7. Confirm that:
   - The warning no longer appears in the PHP error log.
   - Categories with valid thumbnails still render their images as before.
   - Categories whose thumbnail attachment has been removed now render the WooCommerce placeholder image, with no new errors.

<!-- End testing instructions -->

### Testing that has already taken place:
- Tested on a live store running:
  - WordPress 6.8.3
  - WooCommerce (latest release from wordpress.org at the time of testing)
  - Nginx with PHP-FPM on AWS
- Confirmed that:
  - The warning appears when a category thumbnail attachment is missing.
  - The warning no longer appears after applying this patch.
  - Valid thumbnails continue to display correctly.

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Milestone

<!-- DO NOT remove or modify this section (other than to check the box). -->

<!-- milestone-target-selection -->
- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**
<!-- /milestone-target-selection -->

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs created from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message Fix warning when rendering subcategory thumbnails if the thumbnail attachment lookup fails.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

christopher-silvey committed:
Guard array access in woocommerce_subcategory_thumbnail

Guard access to the image URL in woocommerce_subcategory_thumbnail() to avoid
"Trying to access array offset on value of type bool" warnings.

wp_get_attachment_image_src() can return false when the attachment lookup
fails. The previous code assumed the return value was always an array and
unconditionally accessed $image[0], which triggers a warning when it is false.

This change checks that the result is an array and contains index 0 before
reading it. If that check fails, the code now treats the image as unavailable
and allows the existing "if ( $image )" logic later in the function to skip
rendering the thumbnail.

Behavior for valid image data is unchanged.

christopher-silvey committed:
Merge pull request #1 from christopher-silvey/christopher-silvey-patch-1

Guard array access in woocommerce_subcategory_thumbnail

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

christopher-silvey commented:
Thanks for taking a look. 🤜🤛

christopher-silvey committed:
Merge branch 'trunk' into trunk

chihsuan commented:
> Hi @chihsuan ,
> 
> Can you help me get this reviewed?

Thanks for contributing to WooCommerce! @christopher-silvey I've assigned this to the FSE team, as they should be more familiar with this area and can get it reviewed properly.

christopher-silvey commented:
> > Hi @chihsuan ,
> > Can you help me get this reviewed?
> 
> Thanks for contributing to WooCommerce! @christopher-silvey I've assigned this to the FSE team, as they should be more familiar with this area and can get it reviewed properly.

Anytime & thank you so much @chihsuan!

Aljullu commented on plugins/woocommerce/includes/wc-template-functions.php:
```diff
@@ -2841,7 +2841,13 @@ function woocommerce_subcategory_thumbnail( $category ) {
+            if ( is_array( $image ) && isset( $image[0] ) ) {
+            	$image = $image[0];
+            } else {
+            	$image = false;
```
Should we use the placeholder image instead of `false`?

```suggestion
            	$image = wc_placeholder_img_src();
```

This would guarantee the placeholder image is rendered instead of nothing.

christopher-silvey replied:
Anytime and thank you @Aljullu! Awesome, I'm glad you were able to reproduce the issue as well. Thanks for pointing out the linting requirement. Not sure how I missed that placeholder func, so thank you for pointing that out, good eye! I have made these updates :)

Aljullu commented on plugins/woocommerce/includes/wc-template-functions.php:
```diff
@@ -2841,7 +2841,13 @@ function woocommerce_subcategory_thumbnail( $category ) {
+            }
+			
 			$image_srcset = function_exists( 'wp_get_attachment_image_srcset' ) ? wp_get_attachment_image_srcset( $thumbnail_id, $small_thumbnail_size ) : false;
 			$image_sizes  = function_exists( 'wp_get_attachment_image_sizes' ) ? wp_get_attachment_image_sizes( $thumbnail_id, $small_thumbnail_size ) : false;
```
I wonder if we should move the lines that define `$image_srcset` and `$image_sizes` inside the `if`.

Then, inside the `else`, we could do:

```PHP
$image_srcset = false;
$image_sizes  = false;
```

This is in line with what's done when `$thumbnail_id` is undefined. What do you think?

christopher-silvey replied:
I completely agree. I made this update as well 🤝

Aljullu reviewed (commented):
Thanks for opening this PR and good catch with the issue, @christopher-silvey! I could reproduce the issue in a classic theme, setting the _Category display_ option in the customizer (_WooCommerce_ > _Product Catalog_) to _Show subcategories & products_.

I left a couple of questions inline in the code. 👇 

Also, heads-up that there is a [linting issue](https://github.com/woocommerce/woocommerce/actions/runs/19884104069/job/57027990159?pr=62198). The spaces need to be changed to tabs for it to pass.

christopher-silvey committed:
Improve subcategory thumbnail fallback handling

When the category thumbnail attachment lookup fails, fall back to
wc_placeholder_img_src() instead of treating the image as false.

Move the image_srcset and image_sizes assignments into the guarded
branches so they are only set when a valid attachment is found and are
explicitly false when using the placeholder.

This keeps the existing output logic intact while avoiding warnings when
the thumbnail attachment is missing.

christopher-silvey committed:
Merge branch 'trunk' into trunk

christopher-silvey committed:
Remove PHPStan ignore for fixed subcategory thumbnail warning

christopher-silvey committed:
Merge branch 'trunk' into trunk

Aljullu reviewed (approved):
Thanks for the updates, @christopher-silvey, and good job! I tested it again, and it works great. The warning is no longer displayed, and the placeholder is rendered instead:

Before | After
--- | ---
<img width="1051" height="982" alt="Image" src="https://github.com/user-attachments/assets/5a0e4b5e-c19a-48bf-8df8-bdda0a201c4f" /> | <img width="1051" height="982" alt="Image" src="https://github.com/user-attachments/assets/34135bbb-a895-4dab-b8b4-7a1a92489650" />

Code changes look good too, so I'll go ahead and merge it into `trunk`. 🙌  It will likely be included in WooCommerce 10.5, scheduled to be released on February 3rd.

Thanks for your contribution!

<details>
<summary>Testing environment</summary>

* WordPress 6.9
* Plugins: -
* Theme: Storefront
* Store creation: September 2025
* Store size: small
</details>
